src/util/simplify_expr_int.cpp

```diff
@@ -1164,6 +1164,31 @@ bool simplify_exprt::simplify_bitnot(exprt &expr)
   return true;
 }
 
+/// simplifies string equalities as introduced in virtual function removal
+bool simplify_exprt::simplify_string_equality(exprt &expr)
+{
+  if(expr.id() != ID_equal)
+    return true;
+
+  const auto &eq = to_equal_expr(expr);
```

Comment:
should we start using `expr_try_dynamic_cast` instead?

Reply:
What would be the pros/cons of doing so?

Reply:
It would be much cleaner in my opinion.
In particular here `if(expr.id() != ID_equal)` does not check that `expr` has two operands so `to_equal_expr` could potentially fail and abort the program.
The good thing with `expr_try_dynamic_cast` is you don't have to remember what the conditions necessary to be a correct `equal_exprt` are.
With these checks for ID_* on the other hand, if one day we want to change what conditions are necessary to be a correct we would have to go through all `== ID_equal` and `!= ID_equal` and maybe adapt them.

Reply:
I don't see any cons to that, except that it changes from our habits.